python/moosecontrol/runners/interfaces/subprocessrunnerinterface.py

```diff
@@ -0,0 +1,237 @@
+        return self._cleanup_kill_timeout
+
+    def initialize(self):
+        """
+        Initialize the subprocess and optionally the reader thread.
+
+        Should be called first in derived classes within initialize().
+        """
+        if not os.path.isdir(self.directory):
+            raise FileNotFoundError(f"Directory {self.directory} does not exist")
+
+        # Start the process
+        self._process = self.start_process(self.get_full_command(), cwd=self.directory)
+
+        # Start the reader thread if enabled
+        if self.use_subprocess_reader:
+            self._subprocess_reader = SubprocessReader(self._process)
+            self._subprocess_reader.start()
+
+        # Report the PID
+        pid = self.get_pid()
+        assert pid is not None
+        logger.info(f"MOOSE process started with pid {pid}")
+
+    def finalize(self):
+        """
+        Wait for the process and reader thread to end.
+
+        Should be called first in derived classes within finalize().
+        """
+        if self.is_process_running():
+            assert self._process is not None
+            pid = self.get_pid()
+            logger.info(f"Waiting for MOOSE process {pid} to end...")
+            self._process.wait()
+            logger.info("MOOSE process has ended")
+
+        if self._subprocess_reader is not None:
+            if self._subprocess_reader.is_alive():
+                logger.debug("Waiting for the reader thread to end...")
+                self._subprocess_reader.join()
+                logger.debug("Reader thread has ended")
+            self._subprocess_reader = None
+
+    def cleanup(self):
+        """Stop the process and subprocess reader."""
+        if self.is_process_running():
+            logger.info("MOOSE process still running on cleanup")
+
+            # Wait for it to gracefully if we can
+            with TimedPoller(0.001, self.cleanup_kill_timeout) as poller:
+                try:
+                    poller.poll(lambda: not self.is_process_running())
+                except TimedPoller.PollTimeout:
+                    logger.warning(
+                        "MOOSE process still running after "
+                        f"{self.cleanup_kill_timeout:.2f} seconds; killing"
+                    )
+                    self.kill_process()
+                else:
+                    logger.info("MOOSE process ended gracefully")
+
+        if self._subprocess_reader is not None and self._subprocess_reader.is_alive():
+            logger.warning("Reader thread still running on cleanup; waiting")
+            self._subprocess_reader.join()
+            logger.info("Reader thread has ended")
+
+    @abstractmethod
+    def get_additional_command(self) -> list[str]:
+        """
+        Get the additional arguments to start the subprocess with.
+
+        Must be overridden.
+        """
+        raise NotImplementedError  # pragma: no cover
```

Comment:
I don't think it is necessary for this to be abstract. You could just return `[]`. Therefore, the class probably doesn't need to be abstract.

Reply:
It doesn't really make sense to use this class on its own, but I guess

Reply:
Actually yeah, you always need to specify an override to how to connect to the server in the spanwed process. It should always be set. If someone wants to not set it, they can have it really return `[]` in derived if desired.